docs-chatbot/generate-embeddings.ts

```diff
@@ -0,0 +1,105 @@
+import OpenAI from "openai";
+import { promises as fs } from "fs";
+import { join } from "path";
+import dotenv from "dotenv";
+import { encode } from "gpt-tokenizer";
+import * as edgedb from "edgedb";
+import e from "dbschema/edgeql-js";
+import { initOpenAIClient } from "./utils";
+
+dotenv.config({ path: ".env.local" });
+
+const openai = initOpenAIClient();
+
+interface Section {
+  id?: string;
+  path: string;
+  tokens: number;
+  content: string;
+  embedding: number[];
+}
+
+async function walk(dir: string): Promise<string[]> {
+  const immediateFiles = await fs.readdir(dir);
+
+  const recursiveFiles: string[][] = await Promise.all(
+    immediateFiles.map(async (file: any) => {
+      const path = join(dir, file);
+      const stats = await fs.stat(path);
+      if (stats.isDirectory()) return walk(path);
+      else if (stats.isFile()) return [path];
+      else return [];
+    })
+  );
+
+  const flattenedFiles: string[] = recursiveFiles.reduce(
+    (all, folderContents) => all.concat(folderContents),
+    []
+  );
+
+  return flattenedFiles.sort((a, b) => a.localeCompare(b));
+}
+
+async function prepareSectionsData(sectionPaths: string[]): Promise<Section[]> {
+  const contents: string[] = [];
+  const sections: Section[] = [];
+
+  for (const path of sectionPaths) {
+    const content = await fs.readFile(path, "utf8");
+    // OpenAI recommends replacing newlines with spaces for best results (specific to embeddings)
+    const contentTrimmed = content.replace(/\n/g, " ");
+    contents.push(contentTrimmed);
+    sections.push({
+      path,
+      content,
+      tokens: encode(content).length,
+      embedding: [],
+    });
+  }
+
+  const embeddingResponse = await openai.embeddings.create({
+    model: "text-embedding-ada-002",
+    input: contents,
+  });
+
+  embeddingResponse.data.forEach((item, i) => {
+    sections[i].embedding = item.embedding;
+  });
+
+  return sections;
+}
+
+async function generateEmbeddings() {
```

Comment:
This is the main function, so it kinda does everything (it calls all other specific functions). The file name is generate-embeddings.ts so that was my reasoning. @jaclarke proposed we call it updateEmbeddings (which can also be fine, but we actually re-generate and re-store them every time). 

It's also the one that is called inside the IIFE. I think it both generates and stores embeddings. If you strongly believe it should be called storeEmbeddings I will revert it.

Reply:
I don't feel strongly about what it _should_ be called, but I do think it _shouldn't_ be called `generateEmbeddings`. I'm afraid it would be confusing. `updateEmbeddings` seems fine, or we could even just call it `main` and make the IIFE anonymous.